pkg/cmd/cli/cmd/login.go

```diff
@@ -0,0 +1,248 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	kclientcmd "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/cli/config"
+	"github.com/openshift/origin/pkg/cmd/util"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+)
+
+type loginOptions struct {
+	username string
+	password string
+	server   string
+	project  string
+	context  string
+}
+
+func NewCmdLogin(f *osclientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &loginOptions{}
+
+	cmds := &cobra.Command{
+		Use:   "login [--username=<username>] [--password=<password>] [--server=<server>]",
+		Short: "Logs in and returns a session token",
+		Long: `Logs in to the OpenShift server and prints out a session token.
+
+Username and password can be provided through flags, the command will
+prompt for user input if not provided.
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			// fetch config and if a server were not provided, prompt for it
+			clientCfg, err := f.OpenShiftClientConfig.ClientConfig()
+			if isServerNotFound(err) {
+				defaultServer := osclientcmd.DefaultClusterURL
+				promptedServer := util.PromptForStringWithDefault(os.Stdin, defaultServer, "OpenShift is not configured. Please provide the server URL or hit <enter> to use '%v': ", defaultServer)
+				serverFlag := cmd.Flags().Lookup("server")
+				if serverFlag != nil {
+					serverFlag.Value.Set(promptedServer)
+					if err := cmd.Execute(); err != nil {
+						os.Exit(1)
+					} else {
+						os.Exit(0)
+					}
+				}
+			}
+			checkErr(err)
+
+			// try to find a config file from default locations, or create a new one if not found
+			configStore, err := config.GetConfigFromDefaultLocations(clientCfg, cmd)
+			checkErr(err)
+
+			// validate flags
+			checkErr(options.validate(configStore, clientCfg))
+
+			// Check to see if we're already signed in (with the user provided by --username)
+			// If so, simply make sure that .kubeconfig has that information
+			if userFullName, err := whoami(clientCfg); err == nil && (!options.usernameProvided() || (options.usernameProvided() && options.username == userFullName)) {
```

Comment:
----- Original Message -----
> > +			// Check to see if we're already signed in (with the user provided by
> > --username)
> > +			// If so, simply make sure that .kubeconfig has that information
> > +			if userFullName, err := whoami(clientCfg); err == nil &&
> > (!options.usernameProvided() || (options.usernameProvided() &&
> > options.username == userFullName)) {
> 
> Remember that basic auth isn't even our desired end configuration - I'd
> prefer not to store user related info in our config at all.  I don't expect
> users will frequently specify --username and --password (I expect them to
> take them via stdin), but when they do, we should auth exchange for a token
> and replace the existing auth.  

That's actually what we do already - no matter if username and password were provided through flags or stdin, we always exchange for a token and always store the token in the config file. But we also store the username. Here's how it looks like in the file:

```
users:
- name: ffranz
  user:
    token: NzM3ZTk4MWUtNDFjZC00ZmQ5LWJhOGYtYTA5MDJhY2MwZjFk
```

I also don't expect users to provide --username too often, but it's a way to allow you to easily switch between accounts, for example:

```
$ osc login --username=targaryen
Password: *****
Now logged in as 'targaryen'.

$ osc login --username=baratheon
Password: *****
Now logged in as 'baratheon'.
```

> So "if" the current token is invalid, AND if
> the username was specified on the CLI, we should be ignoring the token.  If
> the current token is valid, and --username is provided, we should be
> clearing the current token.

If the current token is valid, and --username is provided, *and the user currently logged in (current context) is the same you wish to login with your provided --username*, why not just keep it logged in? That's what that `if` in the piece of code of the comment above does. In that scenario `login` simply does nothing other than checking the validity of the existing token. 

```
$ osc login --username=baratheon
Password: *****
Now logged in as 'baratheon'.

$ osc login --username=baratheon
Already logged in as 'baratheon'.
```


> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/992/files#r25138884